torch/csrc/jit/serialization/inlined_callstack_serialization.cpp

```diff
@@ -0,0 +1,212 @@
+#include <torch/csrc/jit/api/compilation_unit.h>
+#include <torch/csrc/jit/serialization/inlined_callstack_serialization.h>
+#include <torch/csrc/jit/serialization/pickle.h>
+
+namespace torch {
+namespace jit {
+
+c10::IValue InlinedCallStackSerializer::serialize(
+    const InlinedCallStackPtr& cs_ptr,
+    const SourceRangeTagMap& source_range_tags) {
+  if (!cs_ptr) {
+    return c10::IValue();
+  }
+  auto cs_it = serialized_inlined_callstack_.find(cs_ptr);
+  if (cs_it != serialized_inlined_callstack_.end()) {
+    return cs_it->second;
+  }
+  // Inlined callstack pointer is serialized as tuple of 3 elements
+  // {IValue(module_instance_info), source_range_tag, IValue(InlinedCallStack)}
+  std::vector<c10::IValue> elements;
+  elements.reserve(3);
+  elements.emplace_back(
+      serialize_module_instance_info(cs_ptr->module_instance()));
+  int64_t source_range_tag{-1};
+  const SourceRange& sr = cs_ptr->source_range();
+  auto sr_it = source_range_tags.find(sr);
+  if (sr_it != source_range_tags.end()) {
+    source_range_tag = sr_it->second;
+  }
+  elements.emplace_back(source_range_tag);
+  if (cs_ptr->callee()) {
+    elements.emplace_back(
+        serialize(cs_ptr->callee().value(), source_range_tags));
+  } else {
+    elements.emplace_back(c10::IValue());
+  }
+  c10::IValue serialized_cs = c10::ivalue::Tuple::create(elements);
+  serialized_inlined_callstack_[cs_ptr] = serialized_cs;
+  return serialized_cs;
+}
+
+c10::IValue InlinedCallStackSerializer::serialize_module_instance_info(
+    const c10::optional<ModuleInstanceInfo>& m) {
+  if (m) {
+    const auto& m_val = m.value();
+    std::vector<c10::IValue> elements;
+    // Module instance info is serialized as
+    // {type name, instance name}
+    if (m_val.class_type()) {
+      elements = {
+          m_val.class_type()->name()->qualifiedName(), m_val.instance_name()};
+    } else {
+      elements = {"", m_val.instance_name()};
+    }
+    return c10::ivalue::Tuple::create(std::move(elements));
+  }
+  return c10::IValue();
+}
+
+std::vector<char> InlinedCallStackPickler::pickle(
+    const std::unordered_map<int64_t, DelegateDebugInfoType>& callstack_ptrs,
+    const SourceRangeTagMap& source_range_tags) {
+  std::vector<c10::IValue> ivalues;
+  for (const auto& it : callstack_ptrs) {
+    int64_t debug_handle = it.first;
+    std::vector<c10::IValue> elements;
+    /*
+     * Each debug handle is serialized as a tuple of 3 elements
+     * {debug_handle, source_range_tag, serialized_callstack}
```

Comment:
isn't debug handle the same are source_range_tag? I think I'm getting confused here.

Can you explain a bit more why there's 2 and why they're named differently?

Reply:
So source_range is stored separately in debug_pkl files. Thus when we deserialize source range we create a map of <source_range_tag, source_range>.
Now debug handles point to InlinedCallStack. Since inlined callstack also contain source_range information, in order to populate inlined callstack we need to to know which source range we are referring to. So thats when we use source_range_tag.

I also explained below why there is source_range_tag stored separately, besides the one saved in serialized_callstack.

Not sure if it helps clarify, but if not let me know.

Reply:
Discussed offline. For the specific question, debug_handle has taken a different meaning than before when it was 1on1 with source range tags.

I forgot to ask one last question though, if serilaized_callstack is

{IValue(module_instance_info), source_range_tag, IValue(InlinedCallStack)}

why do we still need source_range_tag in {debug_handle, source_range_tag, serialized_callstack}